internal/locate/region_request.go

```diff
@@ -1395,6 +1397,16 @@ func (s *RegionRequestSender) onRegionError(bo *retry.Backoffer, ctx *RPCContext
 		return true, nil
 	}
 
+	if regionErr.GetRecoveryInProgress() != nil {
+		s.regionCache.InvalidateCachedRegion(ctx.Region)
+		logutil.BgLogger().Debug("tikv reports `RecoveryInProgress`", zap.Stringer("ctx", ctx))
+		err = bo.Backoff(retry.BoRegionRecoveryInProgress, errors.Errorf("region recovery in progress, ctx: %v", ctx))
+		if err != nil {
+			return false, err
+		}
+		return true, nil
```

Comment:
It is a bit strange to both invalidate region cache and return `retry = true`.

If retry is true, in the next retry, it will still use the same `replicaSelector` with the region info in the last time.

So, it depends how the region will change after it get recovered.

If the region epoch or the region range does not change, maybe it's not necessary to invalidate the cache, because a `NotLeader` error will also update the leader.

If the epoch will change, here `retry` should be false. The request will be retried from more upper.

Reply:
The reason to invalidate the cache is that the region epoch in the client cache may be bigger than the recovered region due to lost-updates suffering from majority failure.

Reply:
If you hope that the region cache should refresh in the retry, please return `false, nil` here.